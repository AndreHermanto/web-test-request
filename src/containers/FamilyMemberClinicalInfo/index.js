import React, { Component } from 'react';
import {
  initData,
  setFormData,
  validatedToTrue
} from './reducer'
import { 
  PageHeading,
  FormButton
} from './../../components/SharedStyle';
import TextArea from './../../components/TextArea';
import Toggle from './../../components/Toggle';
import RadioSet from './../../components/RadioSet';

/**
 * FamilyMemberClinicalInfo - UI for input patient details.
 */
class FamilyMemberClinicalInfo extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleBack = this.handleBack.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);
    
    var prefill, prefillClinical;
    if(props.route.data) prefill = props.route.data.familyMembers[props.params.id].familyMemberClinicalInfo;
    if(props.route.clinicalInfoData) prefillClinical = props.route.clinicalInfoData.familyHistory
    this.state = initData(prefill, prefillClinical);
  }
  
  componentDidMount() {
    if(this.props.route.preventUnvisitedFormAccess) {
      this.props.router.setRouteLeaveHook(this.props.route, this.props.route.preventUnvisitedFormAccess.bind(this));
    }
  }
  
  handleChange(event) {
    this.setState(setFormData(this.state, event.target));
  }
  
  handleBack() {
    this.props.route.onChange(this);
    this.props.router.push(`/step4/${this.props.params.mode}/1/${this.props.params.id}`);
  }
  
  handleNext(passValidation) {
    if(!passValidation) return false;
    this.props.route.onChange(this);
    this.props.router.push('/step4');
  }
  
  handleConfirm() {
    return this.setState(validatedToTrue(this.state), () => {    
      var pass = true;
      for (var field in this.state.validation) {
        if(this.state.validation[field].status === 'error') pass = false;
      }
      this.handleNext(pass); 
    });
  }
  
  // This renders the validation result after confirm button is clicked.
  validate() {
    return this.state.validated && this.state.validation;
  }
  
  render() {
    return (
      <div>
        <PageHeading>Step 4-2: {this.props.params.mode} family member {(this.props.params.mode === 'edit') &&  `(${this.props.route.data.familyMembers[this.props.params.id].familyMemberDetails.firstName} ${this.props.route.data.familyMembers[this.props.params.id].familyMemberDetails.lastName})`} - clinical info</PageHeading>
 
        <RadioSet
          label="Is the family member affected?"
          field="affected"
          options={[true, false]}
          formState={this.state.form}
          onChange={this.handleChange}
        />

        <TextArea
          field="clinicalInfo"
          label="Provide Clinical Information"
          helper="Type or copy and paste clinical notes and the diagnosis here. Comprehensive notes increase the chance of a successful diagnosis."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          required={this.state.form.affected}
          optional={!this.state.form.affected}
        />

        <TextArea
          field="relevantInvestigation"
          label="Provide results from relevant investigations"
          helper="Provide notes from genetic tests, imaging results."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          optional
        />
            
        <TextArea
          field="familyHistory"
          label="Family history"
          helper="Provide notes on the suspected inheritance model and affected relatives."
          onChange={this.handleChange}
          onValidate={this.validate()}
          formState={this.state.form}
          optional
        />
            
        <Toggle
          field="consanguinity"
          label="Consanguinity"
          onChange={this.handleChange}
          checked={this.state.form.consanguinity === true}
          formState={this.state.form}
        />

        {this.state.form.consanguinity && (
          <TextArea
            field="consanguinityInfo"
            helper="Please provide details"
            onChange={this.handleChange}
            formState={this.state.form}
          />
        )}
            
        {
          this.props.route.isEdited !== true &&
          <FormButton 
            onClick={this.handleBack}
            label="Back"
            back
          >
            Back
          </FormButton> 
        }
      
        <FormButton 
          type="submit" 
          onClick={this.handleConfirm}
        >
          Next
        </FormButton> 
      </div>
    );
  }
}

export default FamilyMemberClinicalInfo;