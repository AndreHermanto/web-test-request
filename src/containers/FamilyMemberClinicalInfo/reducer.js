import validator from './../../components/validator';

export function initData(prefilled, familyHistory) {
  var state = {
    form: {
      clinicalInfo: '',
      affected: false,
      relevantInvestigation: '',
      familyHistory: '',
      consanguinity: false,
      consanguinityInfo:''
    },
    validationRule: {
      clinicalInfo: 'required'
    },
    validated: false,
    formId: 'familyMemberClinicalInfo'
  };
  
  if(prefilled && Object.keys(prefilled).length !== 0) state.form = prefilled;
  if(familyHistory) state.form.familyHistory = familyHistory;
  
  // Detects affected state to determine whether clinical information state needs to be validated
  state.validationRule.clinicalInfo = state.form.affected ? 'required' : '';
  
  // This validates the data in the initial state.
  state.validation = {};
  for (var field in state.validationRule) {
    if(field) {
      state.validation[field] = validator(state.form[field], state.validationRule[field]);
    }
  }
  
  return Object.assign({}, state);
}


/**
 * This sets the form data upon onChange.
 * @param {Object} state Targeted state to be changed.
 * @param {Object} target Target object captured from UI event change.
 */
export function setFormData(state, target) {
  var value, formStateChild;

  switch(target.type) {
    case 'checkbox':
      value = target.checked;
      break;
    default:
        value = target.value;
      break;
  } 

  if(value === false && target.name === 'consanguinity') {
    formStateChild = Object.assign({}, state.form, {
      [target.name]: value,
      consanguinityInfo: ''
    });
  }
  else {
    formStateChild = Object.assign({}, state.form, {
      [target.name]: value
    });
  }
  
  var validationStateChild = Object.assign({}, state.validation, {
    [target.name]: validator(value, state.validationRule[target.name])
  });

  
  // Affected state will change the clinical info's validation
  if(target.name === 'affected') {
    var newValidationRule = Object.assign({}, state.validationRule, {
      clinicalInfo: value ? 'required' : ''
    });
    
    validationStateChild = Object.assign({}, validationStateChild, {
      clinicalInfo: validator(state.form.clinicalInfo, newValidationRule.clinicalInfo)
    });
    
    return Object.assign({}, state, {
      form: formStateChild,
      validationRule: newValidationRule,
      validation: validationStateChild
    });
  } else {
    return Object.assign({}, state, {
      form: formStateChild,
      validation: validationStateChild
    });
  }
}

/**
 * Set "validated" state to true - identifying the confirm button is clicked and validation processed.
 * @param {Object} state Targeted state to be changed.
 */
export function validatedToTrue(state) {
  return Object.assign({}, state, {
    validated: true
  });
}