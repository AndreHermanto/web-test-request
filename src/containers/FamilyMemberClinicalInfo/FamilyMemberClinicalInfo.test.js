import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TestUtils from 'react-addons-test-utils';
import FetchMock from 'fetch-mock';
import FamilyMemberClinicalInfo from './index';

describe('FamilyMemberClinicalInfo: index', () => {
  var props = {
    route: {
      onChange: jest.fn(),
      onDelete: jest.fn(),
      data: { familyMember: [
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} },
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} },
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} },
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} }
      ] }
    },
    params:{
      mode: 'add',
      id: 3
    },
    router:['/step4/add/2/3']
  };
  
  /*var editProps = {
    route: {
      onChange: jest.fn(),
      onDelete: jest.fn(),
      data: { familyMember: [
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} },
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} },
        { FamilyMemberDetails: {}, FamilyMemberClinicalInfo: {} }
      ] }
    },
    params:{
      mode: 'edit',
      id: 5
    },
    router:['/step4/edit/2/5']
  };*/
  
  test('renders without crashing', () => {
    const page = renderer.create(React.createElement(FamilyMemberClinicalInfo, props)).toJSON();
    expect(page).toMatchSnapshot();
  });
  
  test('handleChange works - general', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));
    const selection = { "target": { 
      "value": "abc",
      "name": "clinicalInfo",
      "type": "text"
    }};
    page.handleChange(selection);
    expect(page.state.form.clinicalInfo).toEqual('abc');  
  });
  
  test('handleChange works - checkbox', () => {
    const page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));
    const selection = { "target": { 
      "checked": true,
      "name": "consangunity",
      "type": "checkbox"
    }};
    page.handleChange(selection);
    expect(page.state.form.consangunity).toEqual(true);  
  });  
  
  test('handleBack works', () =>  {
    var page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));                                         
    page.handleBack();
    expect(page.props.router.pop()).toEqual('/step4/add/1/3')
  });
  
  test('handleCancel works', () =>  {
    var page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));                                         
    page.handleCancel();
    expect(page.props.router.pop()).toEqual('/step4');
  });
  
  test('handleNext works', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));
    page.handleNext(false);
    expect(page.props.router.pop()).toEqual('/step4/add/2/3');
    page.handleNext(true);
    expect(page.props.router.pop()).toEqual('/step4');
  });
  
  test('handleConfirm works', () => {
    var page = TestUtils.renderIntoDocument(React.createElement(FamilyMemberClinicalInfo, props));
    page.handleConfirm();
    expect(page.state.validated).toEqual(true);
  });
});
