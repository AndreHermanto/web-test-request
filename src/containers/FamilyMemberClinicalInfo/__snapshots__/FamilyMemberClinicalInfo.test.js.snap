exports[`FamilyMemberClinicalInfo: index renders without crashing 1`] = `
<div>
  <h3
    className="hWxLNB">
    Step 4: Family Members
  </h3>
  <h4
    className="cZAycE">
    add
     family member 
     - clinical information
  </h4>
  <div
    className="form-group"
    onChange={[Function]}>
    <br />
    <label
      className="radio-inline"
      style={undefined}>
      <input
        defaultChecked={false}
        disabled={false}
        name="affected"
        type="radio"
        value={true} />
      Affected
    </label>
    <label
      className="radio-inline"
      style={undefined}>
      <input
        defaultChecked={true}
        disabled={false}
        name="affected"
        type="radio"
        value={false} />
      Unaffected
    </label>
    <span
      className="SfhHK help-block" />
  </div>
  <div
    className="form-group has-feedback">
    <label
      className="control-label"
      htmlFor={undefined}>
      Provide Clinical Information
      <span
        className="lnMAUb">
        Optional
      </span>
    </label>
    <div
      className="kWhIoX">
      Type or copy and paste clinical notes and the diagnosis here. Comprehensive notes increase the chance of a successful diagnosis.
    </div>
    <textarea
      className="ebbddk form-control"
      id={undefined}
      name="clinicalInfo"
      onChange={[Function]}
      style={
        Object {
          "minHeight": 90,
        }
      }
      type={undefined}
      value="" />
    <span
      className="SfhHK help-block" />
  </div>
  <div
    className="form-group has-feedback">
    <label
      className="control-label"
      htmlFor={undefined}>
      Provide results from relevant investigations
      <span
        className="lnMAUb">
        Optional
      </span>
    </label>
    <div
      className="kWhIoX">
      Provide notes from genetic tests, imaging results.
    </div>
    <textarea
      className="ebbddk form-control"
      id={undefined}
      name="relevantInvestigation"
      onChange={[Function]}
      style={
        Object {
          "minHeight": 90,
        }
      }
      type={undefined}
      value="" />
    <span
      className="SfhHK help-block" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Consanguinity
    </label>
    <br />
    <table>
      <tbody>
        <tr>
          <td>
            <div
              className="react-toggle"
              onClick={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}>
              <div
                className="react-toggle-track">
                <div
                  className="react-toggle-track-check">
                  <svg
                    height="11"
                    viewBox="0 0 14 11"
                    width="14">
                    <title>
                      switch-check
                    </title>
                    <path
                      d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                      fill="#fff"
                      fillRule="evenodd" />
                  </svg>
                </div>
                <div
                  className="react-toggle-track-x">
                  <svg
                    height="10"
                    viewBox="0 0 10 10"
                    width="10">
                    <title>
                      switch-x
                    </title>
                    <path
                      d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                      fill="#fff"
                      fillRule="evenodd" />
                  </svg>
                </div>
              </div>
              <div
                className="react-toggle-thumb" />
              <input
                checked={false}
                className="react-toggle-screenreader-only"
                name="consanguinity"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="checkbox" />
            </div>
          </td>
          <td />
        </tr>
      </tbody>
    </table>
    <span
      className="SfhHK help-block" />
  </div>
  <button
    className="ieDBA"
    label="Back"
    onClick={[Function]}>
    Back
  </button>
  <button
    className="bBdbjl"
    onClick={[Function]}
    type="submit">
    Next
  </button>
</div>
`;
